datafusion/src/row/writer.rs

```diff
@@ -0,0 +1,322 @@
+        assert!(idx < self.field_count);
+    }
+
+    fn set_null_at(&mut self, idx: usize) {
+        let null_bits = &mut self.data[0..self.null_width];
+        set_bit(null_bits, idx, false)
+    }
+
+    fn set_non_null_at(&mut self, idx: usize) {
+        let null_bits = &mut self.data[0..self.null_width];
+        set_bit(null_bits, idx, true)
+    }
+
+    fn set_bool(&mut self, idx: usize, value: bool) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = if value { 1 } else { 0 };
+    }
+
+    fn set_u8(&mut self, idx: usize, value: u8) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = value;
+    }
+
+    fn_set_idx!(u16, 2);
+    fn_set_idx!(u32, 4);
+    fn_set_idx!(u64, 8);
+    fn_set_idx!(i16, 2);
+    fn_set_idx!(i32, 4);
+    fn_set_idx!(i64, 8);
+    fn_set_idx!(f32, 4);
+    fn_set_idx!(f64, 8);
+
+    fn set_i8(&mut self, idx: usize, value: i8) {
+        self.assert_index_valid(idx);
+        let offset = self.field_offsets[idx];
+        self.data[offset] = value.to_le_bytes()[0];
+    }
+
+    fn set_date32(&mut self, idx: usize, value: i32) {
+        set_idx!(4, self, idx, value)
+    }
+
+    fn set_date64(&mut self, idx: usize, value: i64) {
+        set_idx!(8, self, idx, value)
+    }
+
+    fn set_offset_size(&mut self, idx: usize, size: usize) {
+        let offset_and_size: u64 = (self.varlena_offset << 32 | size) as u64;
+        self.set_u64(idx, offset_and_size);
+    }
+
+    fn set_utf8(&mut self, idx: usize, value: &str) {
+        self.assert_index_valid(idx);
+        let bytes = value.as_bytes();
+        let size = bytes.len();
+        self.set_offset_size(idx, size);
+        let varlena_offset = self.varlena_offset;
+        self.data[varlena_offset..varlena_offset + size].copy_from_slice(bytes);
+        self.varlena_offset += size;
+        self.varlena_width += size;
+    }
+
+    fn set_binary(&mut self, idx: usize, value: &[u8]) {
+        self.assert_index_valid(idx);
+        let size = value.len();
+        self.set_offset_size(idx, size);
+        let varlena_offset = self.varlena_offset;
+        self.data[varlena_offset..varlena_offset + size].copy_from_slice(value);
+        self.varlena_offset += size;
+        self.varlena_width += size;
+    }
+
+    fn current_width(&self) -> usize {
+        self.null_width + self.values_width + self.varlena_width
+    }
+
+    /// End each row at 8-byte word boundary.
+    fn end_padding(&mut self) {
+        let payload_width = self.current_width();
+        self.row_width = (payload_width.saturating_add(7) / 8) * 8;
+        if self.data.capacity() < self.row_width {
+            self.data.resize(self.row_width, 0);
+        }
+    }
+
+    fn get_row(&self) -> &[u8] {
+        &self.data[0..self.row_width]
+    }
+}
+
+/// Stitch attributes of tuple in `batch` at `row_idx` and returns the tuple width
+fn write_row(
```

Comment:
I think having a vectorized version of this would also be helpful so an entire record batch could be converted at once